File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/xpath/XPathFacade.java
package com.github.jmlparser.xpath;

import com.github.javaparser.ast.CompilationUnit;
import org.w3c.dom.Document;
import org.w3c.dom.NodeList;

import javax.xml.transform.*;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;
import java.io.StringWriter;

/**
 * @author Alexander Weigl
 * @version 1 (11.02.23)
 */
public class XPathFacade {
    public static String printXml(CompilationUnit n) throws TransformerException {
        Document xmlDocument = DocumentFactories.getDocument(n);
        TransformerFactory transformerFactory = TransformerFactory.newInstance();
        Transformer transformer = transformerFactory.newTransformer();
        transformer.setOutputProperty(OutputKeys.INDENT, "yes");
        Source source = new DOMSource(xmlDocument.getDocumentElement());
        var sw = new StringWriter();
        StreamResult result = new StreamResult(sw);
        transformer.transform(source, result);
        return sw.getBuffer().toString();
    }

    public static void query(CompilationUnit n, String query) throws XPathExpressionException {
        Document xmlDocument = DocumentFactories.getDocument(n);
        XPath xPath = XPathFactory.newInstance().newXPath();
        var res = (NodeList) xPath.compile(query).evaluate(xmlDocument, XPathConstants.NODESET);
        for (int i = 0; i < res.getLength(); i++) {
            System.out.println(((JPElement) res.item(i)).getAstNode());
        }
    }
}

File: tools/xpath/src/main/kotlin/io/github/jmltoolkit/xpath/XPathFacade.kt
package io.github.jmltoolkit.xpath

import com.github.javaparser.ast.CompilationUnit
import org.w3c.dom.Document
import org.w3c.dom.NodeList
import java.io.StringWriter
import javax.xml.transform.OutputKeys
import javax.xml.transform.Source
import javax.xml.transform.TransformerException
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult
import javax.xml.xpath.XPathConstants
import javax.xml.xpath.XPathExpressionException
import javax.xml.xpath.XPathFactory

/**
 * @author Alexander Weigl
 * @version 1 (11.02.23)
 */
object XPathFacade {
    @Throws(TransformerException::class)
    fun printXml(n: CompilationUnit): String {
        val xmlDocument: Document = DocumentFactories.getDocument(n)
        val transformerFactory = TransformerFactory.newInstance()
        val transformer = transformerFactory.newTransformer()
        transformer.setOutputProperty(OutputKeys.INDENT, "yes")
        val source: Source = DOMSource(xmlDocument.documentElement)
        val sw = StringWriter()
        val result = StreamResult(sw)
        transformer.transform(source, result)
        return sw.buffer.toString()
    }

    @Throws(XPathExpressionException::class)
    fun query(n: CompilationUnit, query: String?) {
        val xmlDocument: Document = DocumentFactories.getDocument(n)
        val xPath = XPathFactory.newInstance().newXPath()
        val res = xPath.compile(query).evaluate(xmlDocument, XPathConstants.NODESET) as NodeList
        for (i in 0 until res.length) {
            println((res.item(i) as JPElement).astNode)
        }
    }
}
